### PR TITLE
Add MobilityData GTFS Validator to validation step

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR integrates the [MobilityData GTFS Validator](https://github.com/MobilityData/gtfs-validator) into datatools, and adds a MobilityData validation to the GTFS validation step. Validation results are saved to a unique temporary folder, then loaded and inserted into Mongo. Failures are handled via the existing validation failure infrastructure.

#### Room for Improvement
- [x] I'd like to perhaps see the MobilityData validation happen in a second thread, possibly in parallel to the existing validation. However, I'm not sure how to do this in the cleanest way, and the new validation isn't that slow, especially compared to the existing one.

#### Important Notices
The MobilityData validator requires a much newer version of the Java spec. I've updated by JDK to 18, and language level to 11!

**Tests are failing due to an unrelated shape file export issue**

### CODE REVIEW

**Comment by br648** [comment]
@miles-grant-ibigroup for review: https://github.com/ibi-group/datatools-server/pull/513. We should discuss at some point next week.

**Comment by br648** [comment]
@miles-grant-ibigroup I'm going to defer my review until this https://github.com/ibi-group/datatools-server/pull/513 is tested and merged.

File: pom.xml [hunk 1]
```diff
@@ -163,6 +163,11 @@
             <plugin>
                 <artifactId>maven-surefire-plugin</artifactId>
                 <version>2.22.2</version>
+                <configuration>
+                    <argLine>
+                        --illegal-access=permit
```
Comment (br648) [inline pom.xml:169]: Consider adding a comment to explain this.

File: src/main/java/com/conveyal/datatools/manager/models/transform/NormalizeFieldTransformation.java [hunk 2]
```diff
@@ -234,10 +234,10 @@ public void transform(FeedTransformZipTarget zipTarget, MonitorableJob.Status st
 
             // Copy csv input stream into the zip file, replacing the existing file.
             try (
-                // Modify target zip file that we just read.
-                FileSystem targetZipFs = FileSystems.newFileSystem(tempZipPath, null);
-                // Stream for file copy operation.
-                InputStream inputStream =  new ByteArrayInputStream(stringWriter.toString().getBytes(StandardCharsets.UTF_8))
+                    // Modify target zip file that we just read.
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/transform/NormalizeFieldTransformation.java]: Nit. Restore original indentation.

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 3]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
...
+    public void validateMobility(MonitorableJob.Status status) {
+
+        // Sometimes this method is called when no status object is available.
+        if (status == null) status = new MonitorableJob.Status();
+
+        // VALIDATE GTFS feed
+        FileReader fr = null;
+        try {
+            LOG.info("Beginning MobilityData validation...");
+            status.update("MobilityData Analysis...", 11);
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java:393]: 11% complete, is this just an arbitrary value!

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 4]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
         }
     }
 
+    public void validateMobility(MonitorableJob.Status status) {
+
+        // Sometimes this method is called when no status object is available.
+        if (status == null) status = new MonitorableJob.Status();
+
+        // VALIDATE GTFS feed
+        FileReader fr = null;
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java]: Remove redundant null assignment.

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 5]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
...
+            builder.setGtfsSource(gtfsZip.toURI());
+            builder.setOutputDirectory(Path.of(validatorOutputDirectory));
+            ValidationRunnerConfig mbValidatorConfig = builder.build();
+
+            // Run MobilityData validator
+            ValidationRunner runner = new ValidationRunner(new VersionResolver());
+            runner.run(mbValidatorConfig);
+
+            // Read generated report and save to Mongo
+            fr = new FileReader(validatorOutputDirectory + "report.json");
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java]: Change to try with resources:

`String json;`
`try (FileReader fr = new FileReader(validatorOutputDirectory + "report.json")) {`
`    BufferedReader in = new BufferedReader(fr);`
`    json = in.lines().collect(Collectors.joining(System.lineSeparator()));`
`}`

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java (hunk 4 above)
Comment (br648) [reply src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java]: Remove completely. See try with resources comment below.

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 6]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
...
+        if (status == null) status = new MonitorableJob.Status();
+
+        // VALIDATE GTFS feed
+        FileReader fr = null;
+        try {
+            LOG.info("Beginning MobilityData validation...");
+            status.update("MobilityData Analysis...", 11);
+
+            File gtfsZip = this.retrieveGtfsFile();
+            // Namespace based folders avoid clash for validation being run on multiple versions of a feed
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java]: Nit. Full stop.

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 7]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
...
+            // TODO: do we know that there will always be a namespace?
+            String validatorOutputDirectory = "/tmp/datatools_gtfs/" + this.namespace + "/";
+
+            // Set up MobilityData validator
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java]: Nit. Full stop.

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 8]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
...
+            ValidationRunnerConfig.Builder builder = ValidationRunnerConfig.builder();
+            builder.setGtfsSource(gtfsZip.toURI());
+            builder.setOutputDirectory(Path.of(validatorOutputDirectory));
+            ValidationRunnerConfig mbValidatorConfig = builder.build();
+
+            // Run MobilityData validator
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java:411]: Nit. Full stop.

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 9]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
...
+            ValidationRunner runner = new ValidationRunner(new VersionResolver());
+            runner.run(mbValidatorConfig);
+
+            // Read generated report and save to Mongo
+            fr = new FileReader(validatorOutputDirectory + "report.json");
+            BufferedReader in = new BufferedReader(fr);
+            String json = in.lines().collect(Collectors.joining(System.lineSeparator()));
+            fr.close();
+
+            // This will persist the document to Mongo
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java]: Nit. Full stop.

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 10]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
...
+            ValidationRunnerConfig.Builder builder = ValidationRunnerConfig.builder();
+            builder.setGtfsSource(gtfsZip.toURI());
+            builder.setOutputDirectory(Path.of(validatorOutputDirectory));
+            ValidationRunnerConfig mbValidatorConfig = builder.build();
+
+            // Run MobilityData validator
+            ValidationRunner runner = new ValidationRunner(new VersionResolver());
+            runner.run(mbValidatorConfig);
+
+            // Read generated report and save to Mongo
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java]: Nit. Full stop.

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 11]
```diff
@@ -370,6 +382,48 @@ public void validate(MonitorableJob.Status status) {
         }
     }
 
+    public void validateMobility(MonitorableJob.Status status) {
+
+        // Sometimes this method is called when no status object is available.
+        if (status == null) status = new MonitorableJob.Status();
+
+        // VALIDATE GTFS feed
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java:390]: Nit. Full stop.

**Review by br648 — CHANGES_REQUESTED** [review]

**Comment by miles-grant-ibigroup** [comment]
Thanks Rob! All changes have been addressed

**Review by br648 — APPROVED** [review]
All good. Good luck presenting the new data!

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 12]
```diff
@@ -344,9 +355,10 @@ public void validate(MonitorableJob.Status status) {
         // Sometimes this method is called when no status object is available.
         if (status == null) status = new MonitorableJob.Status();
 
-        // VALIDATE GTFS feed
+        // VALIDATE GTFS feed.
```
Comment (philip-cline) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java:358]: Amazing

**Review by philip-cline — APPROVED** [review]
Looking good to me, seems to work well.

**Pull request merged**